Added Service Discovery for OCI Compute. (#10787)
prometheus/prometheus: docs/configuration/configuration.md

* `__meta_oci_compute_ocpus` : The OCPU count for the instance.
* `__meta_oci_compute_processor_description` :The processor description for the instance.
* `__meta_oci_compute_memory_in_gbs` : The amount of RAM in GBs.
* `__meta_oci_compute_instance_shape` : The OCI compute [shape](https://docs.oracle.com/en-us/iaas/Content/Compute/References/computeshapes.htm) for the instance.
* `__meta_oci_compute_defined_tag_<tagkey>` : Each [defined tag](https://docs.oracle.com/en-us/iaas/Content/Tagging/Concepts/taggingoverview.htm#concepts) value of the instance.
* `__meta_oci_compute_freeform_tag_<tagkey>` : Each [freeform tag](https://docs.oracle.com/en-us/iaas/Content/Tagging/Concepts/understandingfreeformtags.htm) value of the instance.

See below for the configuration options for *OCI Compute* discovery:

```yaml
# The OCID, the unique ID about your OCI tenancy/account. Always required.
tenancy_ocid: <string> 
# The OCI region code, 
# as per https://docs.oracle.com/en-us/iaas/Content/General/Concepts/regions.htm. This determines the OCI API endpoint used by OCI Compute SD. Always required.
region: <string> 
# The OCID for the compartment for your compute instances to be scraped.

> compartment_ocid: <string>

Reviewer: You wonder this could either be a list or omitted maybe?

As to avoid duplicating the whole relabel config in case one has many compartments to scrape?

Author: Hi @neumayer 
We can not omit Compartment, since all OCI metadata API have Compartment-Id as must have field.
All resources in OCI Cloud are organized in Compartment. Think of Compartment as sub-account of AWS, org account in AWS under the same account(I do not exact AWS term).
Yes, we can make it a list, but there is no requirement from the customer as of now, since typically for a job it will be a single Compartment & conceptually another Compartment it will be mostly another job. & if in the future customers want it to be a list, we can update it at that time. Lmk what you think.

Reviewer: If omitted you can query to get a list of all compartments. Or you can specify a list.

The example here is using the node exporter. It is not all that unlikely that users would have instances in multiple compartments. Those would share the same relabelling setup.

I'm not saying it's a must have, and we can work around this at the cost of some copy/paste.

I've been running your sd for a day now and it seems all good, and much easier to use than our previous solution (a script writing the json file for the file based sd). So it would clearly be worth it even with the copy/paste for many compartments (I don't think it's too exotic a use case).

Author: We can update/add support for list of compartments in the future if customers/users ask for it definitely.

[ port: <int> | default = 80 ]
# Refresh interval to re-read the instance list.
[ refresh_interval: <duration> | default = 300s ]

# Optional `authencation` mechanism to be used by OCI Compute SD.
authencation:
  # Sets the authentication type. Permissible enum values: InstancePrincipal or UserPrincipal
  [ type: <enum> | default: InstancePrincipal ]

  # In case UserPrincipal is used, path to your local OCI Config file.
  [ oci_config_file_path: <string> | default: ~/.oci/config]
  # In case UserPrincipal is used, User profile from OCI Config file.
  [ oci_config_file_profile: <string> | DEFAULT ]
  # In case UserPrincipal is used, password for your OCI Config file.
  [ oci_config_file_pvt_key_password: <secret> | default: "" ]

# Tag Filters can be used optionally to filter the instance list, as per the tags applied to them 
# Only instances with 'all' the tags will pass the filter criteria. If instance has more than the following tags, it will pass the filter criteria.
tag_filters:
  [ - tagNamespace: <string> # only needed for defined tags, else can be skipped
      tagKey: <string>
      tagValue: <string>, [...] ]
```

#### OCI IAM policies for authorization 

> When using *InstancePrincipal*, for authentication with OCI APIs, the Prometheus needs to be running on OCI Compute Instance. Addionally compute instance must belong to a dynamic group say *dg_prometheus* with following OCI IAM policies, assigned to the dynamic group. 

Author: will rectify.

[ refresh_interval: <duration> | default = 60s ]
```

### `<oci_compute_sd_config>`

> *OCI Compute* SD configurations allow retrieving scrape targets from [Compute Instances](https://docs.oracle.com/en-us/iaas/Content/Compute/Concepts/computeoverview.htm) from your OCI(Oracle Cloud Infrastructure) tenancy.

Reviewer 2: ```suggestion
OCI Compute SD configurations allow retrieving scrape targets from [Compute Instances](https://docs.oracle.com/en-us/iaas/Content/Compute/Concepts/computeoverview.htm) from your OCI(Oracle Cloud Infrastructure) tenancy.
```

Author: will do

The private IP address is used for scraping by default, but may be changed to the public IP address with relabeling.

The following meta labels are available on targets during [relabeling](#relabel_config):

> * `__meta_oci_compute_instance_id` : The unique ID for the OCI Compute Instance, its OCID.

Reviewer 2: Can you remove the spaces before `:` in the list?

region: <string> 
# The OCID for the compartment for your compute instances to be scraped.
compartment_ocid: <string>
# The port to scrape metrics from. 
[ port: <int> | default = 80 ]
# Refresh interval to re-read the instance list.
[ refresh_interval: <duration> | default = 300s ]

# Optional `authencation` mechanism to be used by OCI Compute SD.
authencation:
  # Sets the authentication type. Permissible enum values: InstancePrincipal or UserPrincipal
  [ type: <enum> | default: InstancePrincipal ]

  # In case UserPrincipal is used, path to your local OCI Config file.
  [ oci_config_file_path: <string> | default: ~/.oci/config]
  # In case UserPrincipal is used, User profile from OCI Config file.
  [ oci_config_file_profile: <string> | DEFAULT ]
  # In case UserPrincipal is used, password for your OCI Config file.
  [ oci_config_file_pvt_key_password: <secret> | default: "" ]

# Tag Filters can be used optionally to filter the instance list, as per the tags applied to them 
# Only instances with 'all' the tags will pass the filter criteria. If instance has more than the following tags, it will pass the filter criteria.
tag_filters:

>   [ - tagNamespace: <string> # only needed for defined tags, else can be skipped

Reviewer 2: ```suggestion
  [ - tag_namespace: <string> # only needed for defined tags, else can be skipped
```

>       tagKey: <string>

Reviewer 2: ```suggestion
      tag_key: <string>
```

>       tagValue: <string>, [...] ]

Reviewer 2: ```suggestion
      tag_value: <string>, [...] ]
```

```

> #### OCI IAM policies for authorization 

Reviewer 2: Add a blank line jere

Author: you mean after "#### OCI IAM policies for authorization" ?

When using *InstancePrincipal*, for authentication with OCI APIs, the Prometheus needs to be running on OCI Compute Instance. Additionally compute instance must belong to a dynamic group say *dg_prometheus* with following OCI IAM policies, assigned to the dynamic group. 

> ```

Reviewer 2: Add an blank line before this

Author: got it before
"#### OCI IAM policies for authorization"

Allow dynamic-group dg_prometheus to read instances in tenancy where target.compartment.id='ocid for the compartment of instances'
Allow dynamic-group dg_prometheus to read vcns in tenancy where target.compartment.id='ocid for the compartment of instances'	
Allow dynamic-group dg_prometheus to read subnets in tenancy where target.compartment.id='ocid for the compartment of instances'	
Allow dynamic-group dg_prometheus to read vnics in tenancy where target.compartment.id='ocid for the compartment of instances'	
Allow dynamic-group dg_prometheus to read vnic-attachments in tenancy where target.compartment.id='ocid for the compartment of instances'	
Allow dynamic-group dg_prometheus to read compartments in tenancy
```

> For detail on *InstancePrincipal*, please refer [documentation for Dynamic groups for OCI Compute Instances](https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm).

Reviewer 2: Add a blank line here

Similaraly, When using *UserPrincipal* for authentication with OCI APIs, the the OCI user say *PrometheusUser*, of the OCI Profile used, needs to have following OCI IAM policies in place.

> ```

Reviewer 2: Again blank line

Allow user PrometheusUser to read instances in tenancy where target.compartment.id='ocid for the compartment of instances'
Allow user PrometheusUser to read vcns in tenancy where target.compartment.id='ocid for the compartment of instances'	
Allow user PrometheusUser to read subnets in tenancy where target.compartment.id='ocid for the compartment of instances'	
Allow user PrometheusUser to read vnics in tenancy where target.compartment.id='ocid for the compartment of instances'	
Allow user PrometheusUser to read vnic-attachments in tenancy where target.compartment.id='ocid for the compartment of instances'	
Allow user PrometheusUser to read compartments in tenancy
```

> For details on setting OCI configuration for *UserPrincipal*, please refer [documentation for OCI-SDK Config setup for users](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/sdkconfig.htm).

Reviewer 2: And a blank line here
